File: cpp/bitdiff.cc
////////////////////////////////////////////////////////////////
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Bit Difference 
 * Hard Accuracy: 46.07% Submissions: 12124 Points: 8
 * We define f(X,Y) as number of different corresponding bits in
 * binary representation of X and Y. For example, f(2,7) = 2,
 * since binary representation of 2 and 7 are 010 and 111,
 * respectively. The first and the third bit differ, so f(2,7)=2.
 *
 * You are given an array A of N integers, A_1,A_2,..,A_N. Find
 * sum of f(A_i,A_j) for all ordered pairs (i,j) such that
 * 1 <= i, j <= N. Return the answer modulo 10**9+7.(!)
 *
 * Example 1:
 * 
 * Input: N = 2
 * A = {2, 4}
 * Output: 4
 * Explaintion: We return 
 * f(2, 2) + f(2, 4) + 
 * f(4, 2) + f(4, 4) = 
 * 0 + 2 + 
 * 2 + 0 = 4.
 *
 * Example 2:
 * 
 * Input: N = 3
 * A = {1, 3, 5}
 * Output: 8
 * Explaination: We return 
 * f(1, 1) + f(1, 3) + f(1, 5) + 
 * f(3, 1) + f(3, 3) + f(3, 5) + 
 * f(5, 1) + f(5, 3) + f(5, 5) = 
 * 0 + 1 + 1 + 
 * 1 + 0 + 2 + 
 * 1 + 2 + 0 = 8.
 * Your Task:
 * You do not need to read input or print anything. Your task is
 * to complete the function countBits() which takes the value N
 * and the array A as input parameters and returns the desired
 * count modulo 10**9+7.
 * Expected Time Complexity: O(N*log2(Max(A_i)))
 * Expected Auxiliary Space: O(1)
 * 
 * Constraints:
 * 1 ≤ N ≤ 10**5
 * 2**0 ≤ A[i] < 2**31
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
////////////////////////////////////////////////////////////////
# include <iostream>
using namespace std;
////////////////////////////////////////////////////////////////
class Solution {
public:
    int countBits( int N, long long int A[]);
};
////////////////////////////////////////////////////////////////
# define GETBIT( j) ((A[j] & mask) > 0)
int Solution::countBits( int N, long long int A[]) {
    static const int NBITS{ 32};
    static const int MODULO{ 1000000007};
    long long cntr{};
    for( int off = 0; off < NBITS; off++) {
        const long long mask = 1 << off;
        int ones = 0;        
        for( int j = 0; j < N; j++) {
            ones += GETBIT( j);
        }
        cntr += 2*ones*( N - ones);
    }
    return( cntr % MODULO);
}
////////////////////////////////////////////////////////////////
int main () {
    int t;
    cin >> t;
    while( t--) {
        int N;
        cin >> N;
        long long int A[N];
        for( int i = 0; i < N; i++) {
            cin >> A[i];
        }        
        Solution ob;
        cout << ob.countBits( N, A) << "\n";
    }
}
////////////////////////////////////////////////////////////////
// log: -Yeah;[Ok], art is horseshit you buy tacos and I'll stop
// posting pixes.
//
// Test Cases Passed: 184 / 184
// Total Time Taken: 0.1/2.3

File: python/bitdiff.py
################################################################
# Sum of bit differences
################################################################
# Stand out from the crowd. Prepare with Complete Interview
# Preparation  
################################################################
# Given an integer array of N integers, find sum of bit differences
# in all pairs that can be formed from array elements. Bit difference
# of a pair (x, y) is count of different bits at same positions in
# binary representations of x and y. For example, bit difference
# for 2 and 7 is 2. Binary representation of 2 is 010 and 7 is 111
# (first and last bits differ in two numbers).
# 
# Note: (x, y) and (y, x) are considered two separate pairs.
################################################################
# Example 1:
# 
# Input: 
# N = 2
# arr[] = {1, 2}
# Output: 4
# Explanation: All pairs in array are (1, 1)
# (1, 2), 2, 1), (2, 2)
# Sum of bit differences = 0 + 2 + 2 + 0
#                        = 4
# Example 2:
# 
# Input:
# N = 3 
# arr[] = {1, 3, 5}
# Output: 8
# Explanation: 
# All pairs in array are (1, 1), (1, 3),
# (1, 5), (3, 1), (3, 3) (3, 5),(5, 1),
# (5, 3), (5, 5)
# Sum of bit differences =  0 + 1 + 1 +
#                           1 + 0 + 2 +
#                           1 + 2 + 0 
#                        = 8
################################################################
# Expected Time Complexity: O(N)
# Expected Auxiliary Space: O(1)
# 
# Constraints:
# 1 <=      N <= 10^5
# 1 <= arr[i] <= 10^5
################################################################
# 2 ^ 16 = 65,536 
# 2 ^ 17 = 131,072
NOOFBITS = 17
class Solution:
    def sumBitDifferences( self, arr, siz ):
        bitsum = [ 0 ] * NOOFBITS
        for n in arr:
            j = 0
            while n > 0:
                bitsum[ j ] += n & 1
                n >>= 1
                j += 1
        total = 0
        for noofones in bitsum:
            noofzeroes = siz - noofones
            total += noofones * noofzeroes * 2
        return total
################################################################
# Test Cases Passed: 360 / 360 
# Total Points Scored: 4 / 4   
# Your Total Score: 966        
# Total Time Taken: 2.49       
# Your Accuracy: 100%          
# Attempts No.: 1              
################################################################
print( Solution().sumBitDifferences([ 5, 7, 2, 4, 8 ], 5 ))
################################################################
# This buffer is for text that is not saved, and for Lisp evaluation.
# To create a file, visit it with C-x C-f and enter text in its buffer.
#
# 2: 0 0 - 0  3: 0 0 0 - 0  4: 0 0 0 0 - 0  5: 0 0 0 0 0 - 5 x 0 = 0
#    0 1 - 1     0 0 1 - 2     0 0 0 1 - 3     0 0 0 0 1 - 4 x 1 = 4
#    1 0 - 1     0 1 0 - 2     0 0 1 0 - 3     0 0 0 1 1 - 3 x 2 = 6
#    1 1 - 0     1 0 0 - 2     0 1 0 0 - 3     0 0 1 1 1 - 2 x 3 = 6
#                0 1 1 - 2     1 0 0 0 - 3     0 1 1 1 1 - 1 x 4 = 4
#                1 0 1 - 2     0 0 1 1 - 4     1 1 1 1 1 - 0 x 5 = 0
#                1 1 0 - 2     0 1 0 1 - 4
#                1 1 1 - 0     0 1 1 0 - 4
#                              1 0 0 1 - 4
#                              1 0 1 0 - 4
#                              1 1 0 0 - 4
#                              0 1 1 1 - 3
#                              1 1 1 0 - 3
#                              1 1 0 1 - 3
#                              1 0 1 1 - 3
#                              1 1 1 1 - 0
# arr[] = {1, 2}      . 3 7              | 8 5 7                       .
# Output: 4           . 1 1 2 x 0 x 2 = 0| 0 1 1  1 x 2 x 2 = 4        .
#                     . 1 1 2 x 0 x 2 = 0| 0 0 1  2 x 1 x 2 = 4        .
# 01                  . 0 1 1 x 1 x 2 = 2| 0 1 1  1 x 2 x 2 = 4        .
# 10                  . -----------------+ 1 0 0  1 x 2 x 2 = 4        .
#                     .     0 + 0 + 2 = 2+---------------------        .
# arr[] = {1, 3, 5}   .                      4 + 4 + 4 + 4 = 16        ,
# Output: 8           . 2 3 4 5 |                                      .
#                     . 0 1 0 1 ' 2 x 2 x 2 = 8                        .
# 001                 . 1 1 0 0 | 2 x 2 x 2 = 8                        .
# 011                 . 0 0 1 1 | 2 x 2 x 2 = 8                        .
# 101                 . --------+--------------                        .
#                     .          8 + 8 + 8 = 24                        .
# 2, 3, 4, 5    => 24 . 5 7 2 4 8 |                                    .
# 3, 7          => 2  . 1 1 0 0 0 | 2 x 3 x 2 = 12                     .
# 8, 5, 7       => 16 . 0 1 1 0 0 ¦ 2 x 3 x 2 = 12                     .
# 5, 7, 2, 4, 8 => 44 . 1 1 0 1 0 | 2 x 3 x 2 = 12                     .
#                     . 0 0 0 0 1 | 4 x 1 x 2 = 8                      .
#                     . ----------+---------------                     .
#                     .      12 + 12 + 12 + 8 = 44                     .
########################################################################
